File: repo/atlas-x402/client/fetch.go
package client

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

type Config struct {
	Network       string
	WalletAddress string
	FacilitatorURL string
	HTTPClient    *http.Client
}

func NewConfig(network, walletAddress string) *Config {
	return &Config{
		Network:       network,
		WalletAddress: walletAddress,
		FacilitatorURL: "https://facilitator.payai.network",
		HTTPClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

func X402Fetch(ctx context.Context, url string, config *Config) (*http.Response, error) {
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, err
	}

	resp, err := config.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode == http.StatusPaymentRequired {
		var paymentReq struct {
			Accepts []struct {
				Scheme           string `json:"scheme"`
				Network          string `json:"network"`
				MaxAmountRequired string `json:"maxAmountRequired"`
				PayTo            string `json:"payTo"`
			} `json:"accepts"`
		}

		if err := json.NewDecoder(resp.Body).Decode(&paymentReq); err != nil {
			resp.Body.Close()
			return nil, err
		}
		resp.Body.Close()

		if len(paymentReq.Accepts) == 0 {
			return nil, fmt.Errorf("no payment requirements provided")
		}

		requirement := paymentReq.Accepts[0]
		paymentPayload := createPayment(requirement, config)

		paymentJSON, err := json.Marshal(paymentPayload)
		if err != nil {
			return nil, err
		}

		paymentHeader := base64.StdEncoding.EncodeToString(paymentJSON)

		req.Header.Set("x-payment", paymentHeader)
		resp, err = config.HTTPClient.Do(req)
		if err != nil {
			return nil, err
		}
	}

	return resp, nil
}

func createPayment(requirement struct {
	Scheme           string `json:"scheme"`
	Network          string `json:"network"`
	MaxAmountRequired string `json:"maxAmountRequired"`
	PayTo            string `json:"payTo"`
}, config *Config) map[string]interface{} {
	return map[string]interface{}{
		"x402Version": 1,
		"scheme":      requirement.Scheme,
		"network":     requirement.Network,
		"payload": map[string]interface{}{
			"transactionHash": "0x...",
			"amount":          requirement.MaxAmountRequired,
			"currency":        "USDC",
			"payTo":           requirement.PayTo,
		},
	}
}

File: repo/atlas-x402/client/fetch.java
package com.atlas402.x402.client;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.util.Base64;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class X402Client {
    private final HttpClient httpClient;
    private final String network;
    private final String walletAddress;
    
    public X402Client(String network, String walletAddress) {
        this.network = network;
        this.walletAddress = walletAddress;
        this.httpClient = HttpClient.newHttpClient();
    }
    
    public CompletableFuture<HttpResponse<String>> fetch(String url) {
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .GET()
            .build();
        
        return httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString())
            .thenCompose(response -> {
                if (response.statusCode() == 402) {
                    return handlePaymentRequired(response, url);
                }
                return CompletableFuture.completedFuture(response);
            });
    }
    
    private CompletableFuture<HttpResponse<String>> handlePaymentRequired(
        HttpResponse<String> response,
        String url
    ) {
        Map<String, Object> paymentReq = parsePaymentRequirements(response.body());
        Map<String, Object> paymentPayload = createPayment(paymentReq);
        
        String paymentHeader = Base64.getEncoder().encodeToString(
            paymentPayload.toString().getBytes()
        );
        
        HttpRequest retryRequest = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .header("x-payment", paymentHeader)
            .GET()
            .build();
        
        return httpClient.sendAsync(retryRequest, HttpResponse.BodyHandlers.ofString());
    }
    
    private Map<String, Object> parsePaymentRequirements(String body) {
        return Map.of();
    }
    
    private Map<String, Object> createPayment(Map<String, Object> requirements) {
        return Map.of(
            "x402Version", 1,
            "scheme", "x402+eip712",
            "network", network,
            "payload", Map.of(
                "transactionHash", "0x...",
                "amount", "1000000",
                "currency", "USDC",
                "payTo", walletAddress
            )
        );
    }
}





